Table expander implementation

Juliakas opened the pull request:
EDIT: In addition to subComponent, I added support for subrows. Subrows are pretty simple to use, just use a `subRows` field in `data` prop object and make a tree-like data structure. For example `const data = [{ ...someData1, subRows: [{..someSubrowsData1}]},{ ...someData2, subRows: [{..someSubrowsData2}]}]`. Padding for nesting has to be provided manually for now.

## Checklist

- [x] Add meaningful tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [x] Add an entry to the changelog for any new components and changes
- [x] Add screenshots of the key elements of the component below
- [x] Add any additional comments below describing the features you've implemented

## Screenshots
### Subcomponent
https://user-images.githubusercontent.com/36437897/117477921-3a53e100-af67-11eb-8e9a-2f04f718c430.mp4
### Subrows demo
https://user-images.githubusercontent.com/36437897/122430183-09ba8a80-cf9c-11eb-95a7-3e96cd7bcf9a.mp4

## Additional Comments
~~This PR is mainly for initiating discussion and evaluating design solutions to match everyone's expectation on how table expanding should work.~~

For me at least, its important to have a flexible way to provide expanding behavior to iTwinUI tables. Solution is inspired by react-table example: https://react-table.tanstack.com/docs/examples/sub-components

- Provided component can be in free, unconstrained format: can be a nested table that fits well, or just some additional information container like in the video example.
- ~~Custom styles can be provided to 'iui-tables-rowgroup' class that wraps both expanded row and expanded content.~~
- Expanding can be invoked freely from Table parent, for example it can be done by providing a column with a button - clicking it will expand a row that is associated with it.

~~Note: there probably should be small stylesheet adjustments, as this selector will break as I have inserted iui-tables-rowgroup div in between~~

veekeys commented on src/core/Table/Table.tsx:
```diff
@@ -86,6 +87,20 @@ export type TableProps<
+   * Component will be placed right after the row, in the same row group div.
+   */
+  expandedSubComponent?: (row: Row<T>) => ReactNode;
+  /**
```
Not sure about the prop name.. But we can discuss this.

mayank99 replied:
Maybe just `subcomponent` (which is what react-table v6 called it) or `subRow`

Juliakas replied:
Renamed to `subComponent`

veekeys replied:
I think expanded prefix I liked better.. That was concrete.. Cause subComponent could be also not necessarily the ex expanded content..Not sure.

bentleyvk replied:
For me `subCompenent` is still not clear without description. Maybe `expansionComponent`, `expandedRow` or `expandedContent.`

bentleyvk replied:
@mayank99 @veekeys can we help Julius with the naming here?

mayank99 replied:
I'm fine with `expandedContent` but I'm also fine with `subcomponent` because that's the term react-table uses

Juliakas replied:
It is currently renamed to subComponent

veekeys commented on src/core/Table/Table.tsx:
```diff
@@ -230,6 +249,7 @@ export const Table = <
     },
     useFlexLayout,
     useSortBy,
+    useExpanded,
```
Just curious..Why decided to put this in the middle? :D

Juliakas replied:
I initially put it as last parameter. But this is the reason, I had to put it in the middle :D
![image](https://user-images.githubusercontent.com/36437897/117635423-7619c100-b188-11eb-9ad9-15e04bdafcf6.png)

veekeys commented on src/core/Table/TableRowMemoized.tsx:
```diff
@@ -42,18 +46,21 @@ const TableRow = <T extends Record<string, unknown>>(props: {
+          );
+        })}
+      </div>
+      {isExpanded && expandedSubComponent?.(row)}
```
I think we might want to arrange spacing ourselves.
![image](https://user-images.githubusercontent.com/81580355/117617201-de5ea780-b174-11eb-9b55-b2792061d584.png)

And somehow I feel that content should start with the 2nd column (I mean not to be aligned with expander column)

@FlyersPh9 Could you check up this PR from visual aspect?

FlyersPh9 replied:
This looks really good, I'm happy you used the borderless button for the expand / collapse arrow.

Question about the row height: is the borderless button causing the rows height to be taller than it would be if it were just text?  If yes, can we adjust the padding so that the row height is close to / the same height as a table row without a button in it?

Aside from that, gets my 👍.

veekeys replied:
@FlyersPh9  And what about padding inside the expanded content? Shouldnt it have some or we leave that freely for the user?

FlyersPh9 replied:
Can you place 1 line of text within the expanded section and try to get the height of that to match the row above?

Juliakas replied:
@FlyersPh9  Yes, it looks like the button is stretching the row a bit since small icon button has fixed height of 27px and the text in other table cells takes up only 22px (due to line-height: 22px). Just by adding cellClassName: 'iui-tables-slot' seems to solve that problem since it removes the paddings from that particular cell. I see the same strategy used for populating checkbox columns.

veekeys commented on stories/core/Table.stories.tsx:
```diff
@@ -307,6 +308,98 @@ Sortable.args = {
+        Header: 'Table',
+        columns: [
+          {
+            id: 'expander',
```
Oh, I think we want to provide this from us. Something like selectable.
Cause this is defined by standards, so we could control that from us, if expanded is enabled (need new prop?). And then check, if we can give users ability to override this if they want.

Juliakas replied:
I moved the expander column to table with latest commit. The behavior can be disabled, if "provideDefaultExpander" prop (true by default) is set to false.

@veekeys What do you think about this?

veekeys replied:
I do not think expand row should be true by default.. Users should turn it on, when they need it.
I like the hook (same as select). Now we need to think, how we could allow this for the user to customize.. I guess same is applicable to selection column.

Juliakas replied:
For me at least, customizing expander from the user side is pretty easy, I only need to add additional Cell at the beggining which invokes `props.row.toggleRowExpanded()` on button click.

I made `provideDefaultExpander` to be `false` by default.

bentleyvk replied:
I think we could provide object/map of predefined columns. It could have expander and maybe even selection. I now think we shouldn't have provided that `isSelectable` flag...

bentleyvk replied:
Maybe instead of `provideDefaultExpander`, we should have a prop for function like `canExpand` or `isExpandable` which would return a boolean whether row should have expansion button.

veekeys commented on stories/core/Table.stories.tsx:
```diff
@@ -307,6 +308,98 @@ Sortable.args = {
+    { name: 'Name2', description: 'Description2' },
+  ],
+};
+
```
Also, I think we will need a story with full features on.
THis way we could test more real life scenarios and also see, how features work together.
For this case, we have a standard, how selected expanded table row looks like, so. need to check that:
![image](https://user-images.githubusercontent.com/81580355/117618373-845ee180-b176-11eb-85e6-4d7db0f574b0.png)

Not necessarily for this PR, we could create a separate one.
@bentleyvk @mayank99

Juliakas replied:
![image](https://user-images.githubusercontent.com/36437897/117657404-2e079800-b1a2-11eb-91a5-b7581e154593.png)
Added another storybook entry for that.

Juliakas replied:
After further experimenting with this, I noticed a couple of issues when combining multiple features. For instance, upon sorting, bottom row reached gets triggered which results in new data being generated, after that sorting simply gets reset. I think this needs to go on a separate PR where fixes could be applied as well.

veekeys replied:
Lets leave this full-on story for separate PR and fixes gonna be there. In this just have a story for expander.

Juliakas commented on src/core/Table/Table.tsx:
```diff
@@ -86,6 +87,20 @@ export type TableProps<
+   * A function that will be used for rendering a component for each row if that row is expanded.
+   * Component will be placed right after the row, in the same row group div.
+   */
+  expandedSubComponent?: (row: Row<T>) => ReactNode;
```
Edited to use React.ReactNode

veekeys commented on src/core/Table/Table.tsx:
```diff
@@ -149,6 +170,9 @@ export const Table = <
     intersectionMargin = 300,
+    expandedSubComponent,
+    onExpand,
+    provideDefaultExpander: provideExpanderColumn = true,
```
New variable looks weird here.. Why you needed it?

Juliakas replied:
Oh.. this is some weird F2 renaming thing in Vscode. When I renamed prop name, it preserved older name and put it after colon... Will fix it

Juliakas replied:
Fixed with new commits

veekeys commented on src/core/Table/Table.tsx:
```diff
@@ -333,6 +392,8 @@ export const Table = <
                 state={state}
                 key={rowProps.key}
+                expandedSubComponent={expandedSubComponent}
+                isExpanded={row.isExpanded}
```
Sure need separate prop `isExpanded`? you can read the value from the row itself inside `TableRowMemoized`.

Juliakas replied:
I previously tried to just use row and would prefer that option, however React.memo prop check for some reason was skipping updates since `prevProp.row.isExpanded` was always equal to `nextProp.row.isExpanded` . I added `isExpanded` separately to fix this issue and now it seems `nextProp.isExpanded` and `prevProp.isExpanded` are actually different.

I'd like more insight on why it is behaving like that, so for now I passed additional prop like that...

veekeys replied:
@bentleyvk thats how it should work?

bentleyvk replied:
We had the same issue with select, you can check like this but with `expanded`:
![image](https://user-images.githubusercontent.com/36186912/118260232-b34cbe80-b4ba-11eb-925d-6981839e49ba.png)

Juliakas replied:
Changed to what you suggested and it does work!

bentleyvk replied:
I guess it is not needed anymore.

Juliakas replied:
Removed it

veekeys commented on src/core/Table/TableRowMemoized.tsx:
```diff
@@ -66,5 +73,7 @@ export const TableRowMemoized = React.memo(
-      nextProp.state.selectedRowIds?.[nextProp.row.id],
+      nextProp.state.selectedRowIds?.[nextProp.row.id] &&
+    prevProp.expandedSubComponent === nextProp.expandedSubComponent &&
+    prevProp.isExpanded === nextProp.isExpanded,
```
Wouldnt this be covered with the check of row.original ?

Juliakas replied:
Same as my above comment, it wasn't covered by just those prop checks

Juliakas commented:
@FlyersPh9 @veekeys

I am currently tweaking some css styling locally and seeing how it behaves with row selection. This is what I came up with so far:
1. I'm thinking about having slightly different backround for subcompoent to distingush from row. Maybe should add even more transparency than it currently has in the video.
2. I adjusted some border css rules - now if row is selected, border will wrap around entire rowgroup. Also added some transparent borders to prevent table from stretching and shrinking on selecting.

### Light
https://user-images.githubusercontent.com/36437897/118250726-6dd6c400-b4af-11eb-9786-7c4fd5186c31.mp4

### Dark
https://user-images.githubusercontent.com/36437897/118250760-76c79580-b4af-11eb-815c-406136202d08.mp4

If all is good, I can make a PR afterwards at iTwinUI repository.

veekeys commented:
> @FlyersPh9 @veekeys
> 
> I am currently tweaking some css styling locally and seeing how it behaves with row selection. This is what I came up with so far:
> 
> 1. I'm thinking about having slightly different backround for subcompoent to distingush from row. Maybe should add even more transparency than it currently has in the video.
> 2. I adjusted some border css rules - now if row is selected, border will wrap around entire rowgroup. Also added some transparent borders to prevent table from stretching and shrinking on selecting.
> 
> Anything else that should be added as well? How about a padding rule for subcomponent? I guess it can be overriden in case user really needs to.
> 
> ### Light
>  Light.mp4 
> ### Dark
>  Dark.mp4 
> If all is good, I can make a PR afterwards at iTwinUI repository.

I like the background!
Border should be around whole section according to standards too. Just not sure about the row border in that case..
Anyway, I think this is already a good base to work on and @FlyersPh9 can improve minor details, where he thinks its needed. You can create PR in iTwinUI repo with this one, if want.

bentleyvk commented on src/core/Table/Table.tsx:
```diff
@@ -373,6 +432,7 @@ export const Table = <
             const rowProps = row.getRowProps({
               className: cx('iui-tables-row', {
                 'iui-tables-row-active': row.isSelected,
+                'iui-tables-expanded': row.isExpanded,
```
Does it do anything?

Juliakas replied:
Yes, just made a PR for styling changes, will be useful for combination of active and expanded rows. And probably for a different background as well. https://github.com/iTwin/iTwinUI/pull/82

bentleyvk commented on src/core/Table/Table.tsx:
```diff
@@ -88,6 +91,24 @@ export type TableProps<
+  /**
+   * Handler for row expand events. Will trigger when expanding and condensing rows.
+   */
+  onExpand?: (
```
Maybe return row data and `newState` only like in other callbacks.

Juliakas commented:
> 
> 
> > @FlyersPh9 @veekeys
> > I am currently tweaking some css styling locally and seeing how it behaves with row selection. This is what I came up with so far:
> > 
> > 1. I'm thinking about having slightly different backround for subcompoent to distingush from row. Maybe should add even more transparency than it currently has in the video.
> > 2. I adjusted some border css rules - now if row is selected, border will wrap around entire rowgroup. Also added some transparent borders to prevent table from stretching and shrinking on selecting.
> > 
> > Anything else that should be added as well? How about a padding rule for subcomponent? I guess it can be overriden in case user really needs to.
> > ### Light
> > Light.mp4
> > ### Dark
> > Dark.mp4
> > If all is good, I can make a PR afterwards at iTwinUI repository.
> 
> I like the background!
> Border should be around whole section according to standards too. Just not sure about the row border in that case..
> Anyway, I think this is already a good base to work on and @FlyersPh9 can improve minor details, where he thinks its needed. You can create PR in iTwinUI repo with this one, if want.

@veekeys 
I'm not sure why I can't directly reply to you on this comment :/ . But anyway, which part of row border exactly? I tried to mimic expanded + selected row from UX standards.

I made a very initial PR for css changes: https://github.com/iTwin/iTwinUI/pull/82
Will fix some failing scenarios now since table structure changes slightly.

bentleyvk commented on src/core/Table/Table.tsx:
```diff
@@ -88,6 +91,24 @@ export type TableProps<
+   * A function that will be used for rendering a component for each row if that row is expanded.
+   * Component will be placed right after the row, in the same row group div.
+   */
+  subComponent?: (row: Row<T>) => React.ReactNode;
```
Not sure whether pass row or row data (`row.original`)

Juliakas replied:
I think `row` could be quite useful since it gives much more information about the row state. I can even do `row.toggleRowExpanded()` inside subcomponent for example.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/iTwin/iTwinUI-react?pullRequest=50) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/iTwin/iTwinUI-react?pullRequest=50) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Julius Rupšys<br/>:x: Juliakas<br/><hr/>**Julius Rupšys** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/iTwin/iTwinUI-react?pullRequest=50) it.</sub>

Juliakas commented:
In this PR I didn't include transition animations - I still need css classes that support trasitions (similar to Toast).

Juliakas commented:
In addition to subComponent, I added support for subrows. Subrows are pretty simple to use, just use a `subRows` field in `data` prop object and make a tree-like data structure. For example `const data = [{ ...someData1, subRows: [{..someSubrowsData1}]},{ ...someData2, subRows: [{..someSubrowsData2}]}]`. Padding for nesting has to be provided manually for now.

bentleyvk commented on src/core/Table/Table.tsx:
```diff
@@ -21,6 +21,7 @@ import {
   useTable,
   ActionType,
   TableInstance,
+  useExpanded,
```
Can you uncomment expand related stuff inside `src\types\react-table-config.ts`

Juliakas replied:
Will do. Just a question, why don't we re-export types from react-table directly?

bentleyvk replied:
Too hard question 😀

bentleyvk commented on src/core/Table/Table.tsx:
```diff
@@ -31,6 +32,8 @@ import SvgSortDown from '@itwin/itwinui-icons-react/cjs/icons/SortDown';
 import { getCellStyle } from './utils';
 import { TableRowMemoized } from './TableRowMemoized';
+import { IconButton } from '../Buttons';
+import { SvgChevronDown, SvgChevronRight } from '@itwin/itwinui-icons-react';
```
Import SVGs like we are importing other SVGs (lines 31, 32). We import like that because other projects does not use tree-shaking and their build will end up with all posible icons.

bentleyvk commented on src/core/Table/Table.tsx:
```diff
@@ -274,6 +323,9 @@ export const Table = <
       case TableActions.setFilter:
         onFilterHandler(newState, action, previousState, instance);
         break;
+      case TableActions.toggleRowExpanded:
```
Add `toggleAllRowsExpanded` too if someone decides to manually make expand all header column.

bentleyvk commented on src/core/Table/Table.tsx:
```diff
@@ -88,6 +91,19 @@ export type TableProps<
+  /**
+   * Handler for row expand events. Will trigger when expanding and condensing rows.
+   */
+  onExpand?: (state: TableState<T>) => void;
```
Add expanded rows data like we do with `onSelect`.

Juliakas replied:
So which data do we want to represent? Should we return row ids for that? For me it would make sense at least.

bentleyvk replied:
I would say expanded rows data (`row.original`) array.

bentleyvk commented on stories/core/Table.stories.tsx:
```diff
@@ -469,6 +472,262 @@ Filters.args = {
+      columns={tableColumnsWithHeaders}
+      data={data}
+      emptyTableContent='No data.'
+      onExpand={(newState) => {
```
You can write jsut `onExpand={onExpand}`.

bentleyvk commented on src/core/Table/Table.tsx:
```diff
@@ -88,6 +92,22 @@ export type TableProps<
+   * Handler for row expand events. Will trigger when expanding and condensing rows.
+   */
+  onExpand?: (
+    selectedData: T[] | undefined,
```
Maybe not `selected` but `expandedData`.

Juliakas replied:
Changed

Juliakas commented:
I have tweaked Creevey tests a bit (at least until the flakiness can be fixed). Most of the tests will ignore images now since it looks like they are not being loaded fast enough for snapshots

veekeys commented:
> I have tweaked Creevey tests a bit (at least until the flakiness can be fixed). Most of the tests will ignore images now since it looks like they are not being loaded fast enough for snapshots

@Juliakas, have you tried using delay? It was enough in Header tests.

Juliakas commented:
> 
> 
> > I have tweaked Creevey tests a bit (at least until the flakiness can be fixed). Most of the tests will ignore images now since it looks like they are not being loaded fast enough for snapshots
> 
> @Juliakas, have you tried using delay? It was enough in Header tests.

I initially thought this would still be flaky. Now I copied delay parameter from Header tests, seems to be working.

veekeys commented:
> > > I have tweaked Creevey tests a bit (at least until the flakiness can be fixed). Most of the tests will ignore images now since it looks like they are not being loaded fast enough for snapshots
> > 
> > 
> > @Juliakas, have you tried using delay? It was enough in Header tests.
> 
> I initially thought this would still be flaky. Now I copied delay parameter from Header tests, seems to be working. I still adjusted Radio button group, the expected image had radio button focused.

It is not something you can trust much, but as for now it seems to work on all the cases...
Radio buttons story is supposed to have focused one. Story shows that you can setFocus on init :D And visual tests test the focus style, so.. How did you make it to be without focus?
https://itwin.github.io/iTwinUI-react/?path=/story/input-inputgroup--radio-group

Juliakas commented:
> 
> 
> > > > I have tweaked Creevey tests a bit (at least until the flakiness can be fixed). Most of the tests will ignore images now since it looks like they are not being loaded fast enough for snapshots
> > > 
> > > 
> > > @Juliakas, have you tried using delay? It was enough in Header tests.
> > 
> > 
> > I initially thought this would still be flaky. Now I copied delay parameter from Header tests, seems to be working. I still adjusted Radio button group, the expected image had radio button focused.
> 
> It is not something you can trust much, but as for now it seems to work on all the cases...
> Radio buttons story is supposed to have focused one. Story shows that you can setFocus on init :D And visual tests test the focus style, so.. How did you make it to be without focus?
> https://itwin.github.io/iTwinUI-react/?path=/story/input-inputgroup--radio-group

In my latest commit I reverted expected image. When running tests it doesn't succeed on the first attempt, I guess this made me think its supposed to be unfocused.

veekeys commented:
> > > > > I have tweaked Creevey tests a bit (at least until the flakiness can be fixed). Most of the tests will ignore images now since it looks like they are not being loaded fast enough for snapshots
> > > > 
> > > > 
> > > > @Juliakas, have you tried using delay? It was enough in Header tests.
> > > 
> > > 
> > > I initially thought this would still be flaky. Now I copied delay parameter from Header tests, seems to be working. I still adjusted Radio button group, the expected image had radio button focused.
> > 
> > 
> > It is not something you can trust much, but as for now it seems to work on all the cases...
> > Radio buttons story is supposed to have focused one. Story shows that you can setFocus on init :D And visual tests test the focus style, so.. How did you make it to be without focus?
> > [itwin.github.io/iTwinUI-react/?path=/story/input-inputgroup--radio-group](https://itwin.github.io/iTwinUI-react/?path=/story/input-inputgroup--radio-group)
> 
> In my latest commit I reverted expected image. When running tests it doesn't succeed on the first attempt, I guess this made me think its supposed to be unfocused.

Ok, now it is weird.. In your PR somehow tests decided to fail, even though on previous PRs they were not failing...
Interesting...

mayank99 commented:
I re-ran the workflow and now it's passing 😆

Juliakas commented:
> 
> 
> > > > > > I have tweaked Creevey tests a bit (at least until the flakiness can be fixed). Most of the tests will ignore images now since it looks like they are not being loaded fast enough for snapshots
> > > > > 
> > > > > 
> > > > > @Juliakas, have you tried using delay? It was enough in Header tests.
> > > > 
> > > > 
> > > > I initially thought this would still be flaky. Now I copied delay parameter from Header tests, seems to be working. I still adjusted Radio button group, the expected image had radio button focused.
> > > 
> > > 
> > > It is not something you can trust much, but as for now it seems to work on all the cases...
> > > Radio buttons story is supposed to have focused one. Story shows that you can setFocus on init :D And visual tests test the focus style, so.. How did you make it to be without focus?
> > > [itwin.github.io/iTwinUI-react/?path=/story/input-inputgroup--radio-group](https://itwin.github.io/iTwinUI-react/?path=/story/input-inputgroup--radio-group)
> > 
> > 
> > In my latest commit I reverted expected image. When running tests it doesn't succeed on the first attempt, I guess this made me think its supposed to be unfocused.
> 
> Ok, now it is weird.. In your PR somehow tests decided to fail, even though on previous PRs they were not failing...
> Interesting...

It is very strange... Because those components that are failing all of the sudden were never touched. Except now when I added 1 second delay for tests only, but they were failing anyway. This time it did succeed though...

mayank99 commented:
> This time it did succeed though...

It didn't at first. So I re-ran the workflow and now it's passing 😆 

I don't think the delay is doing anything.

Juliakas commented:
> 
> 
> > This time it did succeed though...
> 
> It didn't at first. So I re-ran the workflow and now it's passing 😆
> 
> I don't think the delay is doing anything.

Yeah, I meant it like that. Delay for image component seems to work. For Radio group delay probably not..

mayank99 commented on src/core/Table/Table.tsx:
```diff
@@ -88,6 +92,22 @@ export type TableProps<
   intersectionMargin?: number;
+  /**
+   * A function that will be used for rendering a component for each row if that row is expanded.
+   * Component will be placed right after the row, in the same row group div.
```
Update this comment. The row group part is not true anymore.

Also add the part about returning `false` or `null` if you add that functionality.

Juliakas commented on src/core/Table/Table.tsx:
```diff
@@ -190,6 +213,36 @@ export const Table = <
+        minWidth: 48,
+        width: 48,
+        maxWidth: 48,
+        columnClassName: 'iui-tables-slot',
+        cellClassName: 'iui-tables-slot',
```
Changing to iui-slot, removing columnClassName field

Juliakas commented on src/core/Table/Table.tsx:
```diff
@@ -190,6 +213,36 @@ export const Table = <
+        maxWidth: 48,
+        columnClassName: 'iui-tables-slot',
+        cellClassName: 'iui-tables-slot',
+        Cell: ({ row }: CellProps<T>) => (
+          <IconButton
+            styleType='borderless'
+            size='small'
+            onClick={() => {
+              row.toggleRowExpanded();
```
Will look like this: 
![image](https://user-images.githubusercontent.com/36437897/123099054-7c16e900-d43a-11eb-99e1-d5e42ef5ede8.png)

mayank99 replied:
i think checkbox column should go before

Juliakas replied:
My bad, I originally intended it to be before, I think ordering of hooks should fix it

Juliakas commented on src/core/Table/Table.tsx:
```diff
@@ -88,6 +92,22 @@ export type TableProps<
+  ) => void;
+  /**
+   * Flag whether to provide default expander column
+   */
```
Decided to get rid of this prop and use `expanderCell` instead for overriding default

mayank99 commented on .creevey/config.js:
```diff
@@ -17,6 +17,7 @@ module.exports = {
   browsers: {
     chrome: {
       browserName: 'chrome',
+      version: '88.0',
```
Will need to do something about this in a different PR

mayank99 reviewed (approved):
Looks good!
